File: src_8/copula/copula.R
library('copula')

#This function returns copula function
select.copula <- function(cop.name, dimention) {
    if(cop.name == 'frank'){
        function(theta) {
            frankCopula(theta, dim=dimention)
        }
    } else if(cop.name == 'clayton'){
        function(theta) {
            claytonCopula(theta, dim=dimention)
        }
    } else if(cop.name == 'gumbel'){
        function(theta) {
            gumbelCopula(theta, dim=dimention)
        }
    } else if(cop.name == 'normal'){
        function(theta) {
            normalCopula(theta, dim=dimention)
        }
    }
}

#This funtion optimizes the parameter of copula and return it.
# Type of copula (cop.name) and dimention of copula (dimention) are required because range of the parameter is dicided by the type and dimention.
optimize.param <- function(log.likelihood.func, cop.name, dimention) {
    if(cop.name == 'frank'){
        if (dimention == 2) {
            param.upper <- 1000
            param.lower <- -1000
        } else {
            param.upper <- 1000
            param.lower <- 0
        }
    } else if(cop.name == 'clayton'){
        if (dimention == 2) {
            param.upper <- 1000
            param.lower <- -1
        } else {
            param.upper <- 1000
            param.lower <- 0
        }
    } else if(cop.name == 'gumbel'){
        param.upper <- 1000
        param.lower <- 1
    } else if(cop.name == 'normal'){
        param.upper <- 1
        param.lower <- -1
    }
    result <- optimize(f=log.likelihood.func, c(param.lower,param.upper), maximum=TRUE)
    return(result$maximum)
}

#This constructs optimized copula from training data. You have to choice the type of copula(frank, clayton, gumbel or normal).
copula.constructor <- function(training.marginal.dist.matrix, cop.name, param, dim) {
    dimention <- ncol(training.marginal.dist.matrix)
    if (!is.null(dim)) {
        dimention <- dim
    }
    selected.copula <- select.copula(cop.name, dimention)
    if (!is.null(param)) {
        optimized.cop <- selected.copula(param)
        return (optimized.cop)
    }
    log.density.func <- function(cop) {
        log(pCopula(training.marginal.dist.matrix, cop))
    }
    #Create log likelihood func for optimizing parameter
    log.likelihood.func <- function(theta) {
        cop <- selected.copula(theta)
        log.density <- log.density.func(cop)
        L <- 1 + sum(sapply(log.density,
        function(d) { if (is.na(d)) return (0) else if (d == Inf) return (1000) else if (d == -Inf) return (-1000) else return (d)}))
        return (L)
    }
    optimized.param <- optimize.param(log.likelihood.func, cop.name, dimention)
    optimized.cop <- selected.copula(optimized.param)
}

if (py.cop.name == 'indep') {
    trained <- indepCopula(dim=ncol(training.marginal.dist.matrix))
    trained.param <- 'indep'
} else {
    trained <- copula.constructor(py.training.marginal.dist.matrix, py.cop.name, py.param, py.dim)
    trained.param <- trained@parameters[1]
}

File: src_8/copula/copula.py
import pyper
import math
import numpy as np


# Wrapper class of copula implemented in R.
class Copula:
    # if 'param' is None, the parameter of copula is estimated in R.
    def __init__(self, training_marginal_dist_matrix: np.matrix, family: str, param=None, dim=None):
        if family not in ['gumbel', 'clayton', 'frank', 'normal', 'indep']:
            print('Copula family "' + family + '" is not supported.')
            raise ValueError
        if training_marginal_dist_matrix.size == 0:
            if dim is None:
                raise ValueError
            self._dimension = dim
        else:
            self._dimension = training_marginal_dist_matrix.shape[1]
        self._r_engine = pyper.R()
        self._r_engine.assign("py.training.marginal.dist.matrix", training_marginal_dist_matrix)
        self._r_engine.assign("py.cop.name", family)
        self._r_engine.assign("py.param", param)
        self._r_engine.assign("py.dim", self._dimension)
        self._r_engine('source("copula/copula.R")')
        self._r_engine('source("copula/copula.R")')
        trained_param = self._r_engine.get("trained.param")
        self.trained_param = trained_param#asahi
        if trained_param is None:
            self._r_engine('trained <- indepCopula(dim=%d)' % self._dimension)
            print('indep')
        else:
            print(trained_param)

    # Caution! This method is not thread safe.
    def cdf(self, marginal_dist_matrix: np.matrix) -> float:
        if self._dimension != marginal_dist_matrix.shape[1]:
            return None
        self._r_engine.assign("py.marginal.dist.matrix", marginal_dist_matrix)
        cdf = self._r_engine.get("pCopula(py.marginal.dist.matrix, trained)")
        if cdf is None:
            return 0
        if math.isnan(cdf):
            return 0
        return cdf

    # Caution! This method is not thread safe.
    def pdf(self, marginal_dist_matrix: np.matrix) -> float:
        if self._dimension != marginal_dist_matrix.shape[1]:
            return None
        self._r_engine.assign("py.marginal.dist.matrix", marginal_dist_matrix)
        pdf = self._r_engine.get("dCopula(py.marginal.dist.matrix, trained)")
        if pdf is None:
            return 0
        if math.isnan(pdf):
            return 0
        return pdf

    def get_optimized_param(self) -> float:
        return self._r_engine.get("trained.param")

    def get_param(self)->dict:
        return {'trained_param':self.trained_param,'dimension':self._dimension}
